src/worktree.c

```diff
@@ -0,0 +1,231 @@
+#include "utils.h"
+
+// Note that `valid` and `locked` only report `TRUE` if the column is known to
+// be valid or locked (i.e. unknowable due to an error reports `FALSE`, like
+// their `git_worktree_is_valid()` and `git_worktree_is_locked()` counterparts).
+SEXP R_git_worktree_list(SEXP ptr) {
+  git_strarray worktrees = {0};
+  git_repository *repo = get_git_repository(ptr);
+  bail_if(git_worktree_list(&worktrees, repo), "git_worktree_list");
+  SEXP names = PROTECT(Rf_allocVector(STRSXP, worktrees.count));
+  SEXP paths = PROTECT(Rf_allocVector(STRSXP, worktrees.count));
+  SEXP valid = PROTECT(Rf_allocVector(LGLSXP, worktrees.count));
+  SEXP locked = PROTECT(Rf_allocVector(LGLSXP, worktrees.count));
+  for (size_t i = 0; i < worktrees.count; i++) {
+    git_worktree *worktree = NULL;
+    char* name = worktrees.strings[i];
+    SET_STRING_ELT(names, i, safe_char(name));
+    if (git_worktree_lookup(&worktree, repo, name) == GIT_OK) {
+      SET_STRING_ELT(paths, i, safe_char(git_worktree_path(worktree)));
+      SET_LOGICAL_ELT(valid, i, git_worktree_validate(worktree) == GIT_OK);
+      SET_LOGICAL_ELT(locked, i, git_worktree_is_locked(NULL, worktree) > 0);
+      git_worktree_free(worktree);
+    } else {
+      SET_STRING_ELT(paths, i, NA_STRING);
+      SET_LOGICAL_ELT(valid, i, NA_LOGICAL);
+      SET_LOGICAL_ELT(locked, i, NA_LOGICAL);
+    }
+    free(name);
+  }
+  SEXP out = build_tibble(
+    4,
+    "name", names,
+    "path", paths,
+    "valid", valid,
+    "locked", locked
+  );
+  UNPROTECT(4);
+  return out;
+}
+
+SEXP R_git_worktree_exists(SEXP ptr, SEXP name) {
+  git_repository *repo = get_git_repository(ptr);
+  git_worktree *worktree = NULL;
+  const int exists = git_worktree_lookup(&worktree, repo, CHAR(STRING_ELT(name, 0))) == GIT_OK;
+  git_worktree_free(worktree);
+  return Rf_ScalarLogical(exists);
+}
+
+SEXP R_git_worktree_path(SEXP ptr, SEXP name) {
+  git_repository *repo = get_git_repository(ptr);
+  git_worktree *worktree = NULL;
+  bail_if(git_worktree_lookup(&worktree, repo, CHAR(STRING_ELT(name, 0))), "git_worktree_lookup");
+  SEXP out = PROTECT(safe_string(git_worktree_path(worktree)));
+  git_worktree_free(worktree);
+  UNPROTECT(1);
+  return out;
+}
+
+// Throws away reason for why it is not valid. Could add
+// `git_worktree_check_valid()` to propagate the reason up.
+SEXP R_git_worktree_is_valid(SEXP ptr, SEXP name) {
```

Comment:
A few functions like `R_git_worktree_is_valid()` currently throw away a bit of info to be more R-like.

For example, the git function `git_worktree_validate()` returns 0 or an error code. We just return `TRUE` or `FALSE` so we lose the error information here. 

An R level `git_worktree_check_valid()` would retain that error info as part of an error message, but I don't think it is worth it right now.